### PR TITLE
CHI-1385 lambda deployment

### PR DESCRIPTION
**Pull request by robert-bo-davis**
## Description
The basic lambda deploy code is tested using a push event and is ready for merge into master. It shouldn't impact anything outside of lambdas so should be safe to merge.

We are depending on this for some new features.

### Checklist
- [x] Corresponding issue has been opened

### Related Issues
CHI-1385

### CODE REVIEW

File: .github/workflows/hrm-lambdas-release.yml
Comment (GPaoloni) [inline .github/workflows/hrm-lambdas-release.yml:70]: I'm not sure I get this chunk, what's the intention?

File: .github/workflows/htm-all-lambdas-release.yml
Comment (GPaoloni) [inline .github/workflows/htm-all-lambdas-release.yml:11]: Probably not needed here?

File: .github/workflows/hrm-lambdas-release.yml
Comment (GPaoloni) [inline .github/workflows/hrm-lambdas-release.yml:21]: Maybe I'm too paranoid, but having to type this was a security mechanism we introduced to avoid unintended deployments that can be triggered by clicking (like in hrm). Not really asking to change but it would be nice to know what the rest of the team thinks.

Comment (robert-bo-davis) [reply .github/workflows/hrm-lambdas-release.yml:21]: That makes sense to me. I'll definitely defer to the team.

I will say that 99% of the time I use git tags to trigger deployments and make everything "smart" enough to figure out what it has to do. This has 4 main benefits:

1. It lets you know by looking at git exactly what code is deployed where.
2. It allows for easy rollbacks without having to revert changes in master. Want to roll back? checkout the correct commit and create a new tag.
3. you can deploy from non-master branches (especially useful for things like dynamic dev envs), but also nice for emergency hotfixes or publicly testing long running feature branches.
4. It makes automation tasks like slack ops fairly simple.

example:

```yaml
  on:
    push:
      tags: ['dynamic-*', development, staging, canary, production]
```


Generally i would set up a script in package.lock or makefile for non JS projects so a user just has to do:

`npm run deploy:development` or `npm run deploy:staging` and the script handles updating the git tag appropriately. I also usually add a verification step to the script for production deployments as a safety catch. So dev, staging, etc deploy immediately since they are most common and not likely to break anything user facing. Production asks a "you are deploying production, are you sure?" prompt before continuing.

File: .github/workflows/htm-all-lambdas-release.yml
Comment (robert-bo-davis) [reply .github/workflows/htm-all-lambdas-release.yml:11]: Good catch. Already have it fixed in my local repo, just haven't pushed yet.

File: .github/workflows/hrm-lambdas-release.yml
Comment (robert-bo-davis) [reply .github/workflows/hrm-lambdas-release.yml:70]: This is an example of setting the env var AWS_REGION based on the environment instead of having to type/select two things. I still don't fully understand the region/env setup and could use some pointers on that. If an environment exists across regions, this could set an output to be used in a matrix workflow later to deploy to all regions for that environment without the dev knowing what they all are.

Comment (GPaoloni) [reply .github/workflows/hrm-lambdas-release.yml:70]: Ah okay, I get it now. About the region-environment, our setup is like:
- region 1:
  - development
  - staging
  - production
- region 2:
  - development
  - staging
  - production

and so on. It would be ideal to preserve the ability to deploy to a particular region + environment combination, but having a "wrapper action" that "deploys to all staging" or something like that is also desirable in the short term.

Comment (robert-bo-davis) [reply .github/workflows/hrm-lambdas-release.yml:70]: That makes sense. Right now I can only see a staging env in eu-west-1, but that may be a permissions issue.

Out of curiosity, what is a use case where one would want to deploy to development/staging/production in only one of the two regions?

Comment (GPaoloni) [reply .github/workflows/hrm-lambdas-release.yml:70]: Just mentioned that we may want to get rid of that granularity in https://github.com/techmatters/infrastructure-config/pull/78#issuecomment-1259844008 :p

File: .github/workflows/hrm-lambda-release.yml
Comment (robert-bo-davis) [inline .github/workflows/hrm-lambda-release.yml:118]: I had to add these entries to arn:aws:iam::712893914485:policy/TL_GitHub_ECR_ECS_Policy so that this step would work:

```

                "arn:aws:ecr:eu-west-1:712893914485:repository/development/*",
                "arn:aws:ecr:eu-west-1:712893914485:repository/staging/*",
                "arn:aws:ecr:eu-west-1:712893914485:repository/production/*"
 ```
 
 This demos how namespacing resources can simplify IAM permissions as well. I can change these to be specific to the exact repos, if you would prefer.

Comment (robert-bo-davis) [reply .github/workflows/hrm-lambda-release.yml:118]: Let me know if that policy is managed in TF somewhere. I couldn't find it anywhere, but am happy to update it.

**Comment by robert-bo-davis** [comment]
The action to deploy a single lambda is tested via push. I'll remove the push part and maybe simplify the env hack once the basic flow is reviewed and approved.

File: .github/workflows/hrm-lambdas-release.yml
Comment (stephenhand) [reply .github/workflows/hrm-lambdas-release.yml]: We currently only have a staging HRM enviroment in eu-west-1, a prod environment is planned - we don't have any production helplines in that region yet though

File: .github/workflows/hrm-ecs-ci-multiregion.yml
Comment (stephenhand) [inline .github/workflows/hrm-ecs-ci-multiregion.yml:25]: This can probably be renamed to simply 'Deploy HRM to Amazon ECS' - the [Multi Region] qualifier was only there for when we still had the legacy workflow with a hardcoded region availalbe

**Review by stephenhand — APPROVED** [review]
This looks great, couple of super minor comments

File: .github/workflows/hrm-ecs-ci-multiregion.yml
Comment (robert-bo-davis) [reply .github/workflows/hrm-ecs-ci-multiregion.yml:25]: I haven't actually made any changes to the hrm deployment yet. This was my IDE applying it's yaml lint rules on save (oops). I can revert these changes if needed.

File: .github/workflows/hrm-lambda-release.yml
Comment (GPaoloni) [inline .github/workflows/hrm-lambda-release.yml:28]: I think in `workflow_dispatch` we always want to send a Slack message, and maybe not to do so on `workflow_call` event (reusable workflows)?

File: .github/workflows/hrm-lambda-release.yml
Comment (GPaoloni) [inline .github/workflows/hrm-lambda-release.yml:98]: This might be a dumb question, but what the `:live` suffix stands for? Is that something we may want to specify as input?

File: .github/workflows/hrm-lambda-release.yml
Comment (GPaoloni) [inline .github/workflows/hrm-lambda-release.yml:102]: If this assumes that the function to be deployed will be prefixed with the `${{ env.ENVIRONMENT }}-hrm-contact-job-`, I would prefer to rename this workflow to something like "Deploy hrm-contact-job Lambda". The other option could be remove that hard coded chunk and rely on the lambdas source being called after `hrm-contact-job-xxxxxx`.

**Review by GPaoloni — APPROVED** [review]
Looks good to me! Leaving some comments but don't need to change anything for now, just raising more discussions for the next iterations.

File: .github/workflows/hrm-lambda-release.yml
Comment (robert-bo-davis) [reply .github/workflows/hrm-lambda-release.yml:102]: Namespacing the lambda function allows for simplified IAM permissions like:

```
{
   "Version": "2012-10-17",
   "Statement": [
       {
           "Sid": "VisualEditor0",
           "Effect": "Allow",
           "Action": [
               "lambda:UpdateFunctionCode"
           ],
           "Resource": [
               "arn:aws:lambda:us-east-1:712893914485:function:development-*",
               "arn:aws:lambda:us-east-1:712893914485:function:staging-*",
               "arn:aws:lambda:us-east-1:712893914485:function:production-*"
           ]
       }
   ]
}
```

I'd strongly suggest we move towards this style of aws naming. For ARNS that support `/`, I would generally recommend names like `development/some-name` so you can apply env based IAM like;

```
"arn:aws:ecr:us-east-1:712893914485:repository/development/*",
```

Comment (robert-bo-davis) [reply .github/workflows/hrm-lambda-release.yml:102]: You can get more granular with the permissions if someone (or some service) only need access to hrm functions in development you can just give them access to: `arn:aws:lambda:us-east-1:712893914485:function:development-hrm-*` if they/it just needs access to contact jobs in hrm in development you can do `arn:aws:lambda:us-east-1:712893914485:function:development-hrm-contact-jobs-*`.

Using naming conventions to extend IAM permissions reduces the need to modify a policy every time a developer adds something new to the stack... which gets VERY cumbersome as infrastructures grow more complex. The order of the naming becomes very important when you start thinking this way about permissions.

File: .github/workflows/hrm-lambda-release.yml
Comment (robert-bo-davis) [reply .github/workflows/hrm-lambda-release.yml:98]: In the current setup where every environment will have an ECR repo in every region, it is just a hardcoded tag for "this is the docker image tag that is currently running in this enviornment". I'm not sure I can see a use case for changing it on the fly.

In the setup I'd suggest moving to at some point, where there is one public repo for each lambda type, the substitution will change. It will be something like: `public.ecr.aws/techmatters/hrm-lambdas-${{ env.LAMBDA_NAME }}:${{ env.ENVIRONMENT }}` so you can look at docker tags and see what image is running in each environment.

File: .github/workflows/hrm-lambda-release.yml
Comment (robert-bo-davis) [reply .github/workflows/hrm-lambda-release.yml:28]: Ahh! I was confused about the workflow call. I'll updated this. Thanks.

File: .github/workflows/hrm-lambda-release.yml
Comment (robert-bo-davis) [reply .github/workflows/hrm-lambda-release.yml:59]: It could become less complicated since push triggers are disabled, but for simplicity of testing, I'd like to leave it in until things are a bit more settled.

**Pull request merged**
